File: clients/csharp/Runtime/Core/Metrics/HyperLogLog.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Text;

namespace Wardx
{
    public interface IDistinct
    {
        void Add(string identifier);
    }

    public sealed class HllBody
    {
        public readonly int Precision;
        public readonly string Registers;

        public HllBody(int precision, string registers)
        {
            Precision = precision;
            Registers = registers;
        }
    }

    public sealed class HyperLogLog : IDistinct
    {
        public const int Precision = 9;
        public const int RegisterCount = 1 << Precision;

        readonly string _privacySalt;
        readonly byte[] _registers = new byte[RegisterCount];

        public readonly string Name;
        public readonly IReadOnlyDictionary<string, object> Dims;
        public bool Dirty { get; private set; }

        public HyperLogLog(string name, IReadOnlyDictionary<string, object> dims, string privacySalt)
        {
            if (string.IsNullOrEmpty(privacySalt))
            {
                throw new ArgumentException("distinct requires a non-empty privacySalt");
            }
            Name = name;
            Dims = dims;
            _privacySalt = privacySalt;
        }

        public void Add(string identifier)
        {
            if (string.IsNullOrEmpty(identifier))
            {
                throw new ArgumentException("distinct.add requires a non-empty string");
            }
            byte[] digest;
            using (var sha = SHA256.Create())
            {
                digest = sha.ComputeHash(Encoding.UTF8.GetBytes(_privacySalt + "\0" + identifier));
            }
            var index = (digest[0] << 1) | (digest[1] >> 7);
            var rank = RankAfterIndex(digest);
            if (rank > _registers[index]) _registers[index] = rank;
            Dirty = true;
        }

        public HllBody Snapshot()
        {
            return new HllBody(Precision, Convert.ToBase64String(_registers));
        }

        public void Reset()
        {
            Array.Clear(_registers, 0, _registers.Length);
            Dirty = false;
        }

        static byte RankAfterIndex(byte[] digest)
        {
            byte rank = 1;
            for (var bit = Precision; bit < 64; bit++)
            {
                if ((digest[bit >> 3] & (1 << (7 - (bit & 7)))) != 0) return rank;
                rank++;
            }
            return rank;
        }
    }

    public sealed class NoopDistinct : IDistinct
    {
        public static readonly NoopDistinct Instance = new NoopDistinct();
        public void Add(string identifier) { }
    }
}

File: packages/core/src/metrics/HyperLogLog.js
import { createHash } from 'node:crypto';

export const HLL_PRECISION = 9;
export const HLL_REGISTER_COUNT = 1 << HLL_PRECISION;
export const HLL_MAX_RANK = 64 - HLL_PRECISION + 1;

function assertRegisters(registers) {
  if (!(registers instanceof Uint8Array) || registers.length !== HLL_REGISTER_COUNT) {
    throw new Error(`HLL registers must contain exactly ${HLL_REGISTER_COUNT} bytes`);
  }
  for (const rank of registers) {
    if (rank > HLL_MAX_RANK) throw new Error(`HLL register rank must be <= ${HLL_MAX_RANK}`);
  }
}

function rankAfterIndex(digest) {
  let rank = 1;
  for (let bit = HLL_PRECISION; bit < 64; bit++) {
    if ((digest[bit >> 3] & (1 << (7 - (bit & 7)))) !== 0) return rank;
    rank += 1;
  }
  return rank;
}

export function encodeHllRegisters(registers) {
  assertRegisters(registers);
  return Buffer.from(registers).toString('base64');
}

export function decodeHllRegisters(body) {
  if (!body || body.precision !== HLL_PRECISION || typeof body.registers !== 'string') {
    throw new Error(`HLL sketch must use precision ${HLL_PRECISION}`);
  }
  const decoded = Buffer.from(body.registers, 'base64');
  if (decoded.toString('base64') !== body.registers) throw new Error('HLL registers must be canonical base64');
  const registers = Uint8Array.from(decoded);
  assertRegisters(registers);
  return registers;
}

export function estimateHllRegisters(registers) {
  assertRegisters(registers);
  let harmonic = 0;
  let zeros = 0;
  for (const rank of registers) {
    harmonic += 2 ** -rank;
    if (rank === 0) zeros += 1;
  }
  const m = HLL_REGISTER_COUNT;
  const alpha = 0.7213 / (1 + 1.079 / m);
  const raw = alpha * m * m / harmonic;
  const corrected = raw <= 2.5 * m && zeros > 0 ? m * Math.log(m / zeros) : raw;
  return Math.round(corrected);
}

export function estimateHyperLogLog(body) {
  return estimateHllRegisters(decodeHllRegisters(body));
}

export function mergeHyperLogLog(left, right) {
  const leftRegisters = decodeHllRegisters(left);
  const rightRegisters = decodeHllRegisters(right);
  for (let index = 0; index < leftRegisters.length; index++) {
    if (rightRegisters[index] > leftRegisters[index]) leftRegisters[index] = rightRegisters[index];
  }
  return { precision: HLL_PRECISION, registers: encodeHllRegisters(leftRegisters) };
}

export class HyperLogLog {
  constructor(name, dims, privacySalt) {
    if (typeof privacySalt !== 'string' || privacySalt.length === 0) {
      throw new Error('distinct requires a non-empty privacySalt');
    }
    this.name = name;
    this.dims = dims;
    this.privacySalt = privacySalt;
    this.registers = new Uint8Array(HLL_REGISTER_COUNT);
    this.dirty = false;
  }

  add(identifier) {
    if (typeof identifier !== 'string' || identifier.length === 0) {
      throw new Error('distinct.add requires a non-empty string');
    }
    const digest = createHash('sha256')
      .update(this.privacySalt, 'utf8')
      .update('\0', 'utf8')
      .update(identifier, 'utf8')
      .digest();
    const index = (digest[0] << 1) | (digest[1] >> 7);
    const rank = rankAfterIndex(digest);
    if (rank > this.registers[index]) this.registers[index] = rank;
    this.dirty = true;
  }

  snapshot() {
    return { precision: HLL_PRECISION, registers: encodeHllRegisters(this.registers) };
  }

  reset() {
    this.registers.fill(0);
    this.dirty = false;
  }
}

export const NOOP_DISTINCT = Object.freeze({ add() {} });
